[Bug] Missing Generic type parameters on AnyReceiptEnvelope implementation

### Component

consensus, eips, genesis, provider, pubsub

### What version of Alloy are you on?

rev = "fe5754c"

### Operating System

Linux

### Describe the bug

In the code below:

```rust
use std::env;

use alloy_primitives::{address, b256, utils::format_ether, Address, B256};
use alloy_provider::{network::AnyNetwork, Provider, ProviderBuilder};
use alloy_sol_types::{sol, SolEvent};
use dotenv::dotenv;
use eyre::Result;
use IERC20::Transfer;

const WETH: Address = address!("C02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2");

// Transfer topic:
// https://etherscan.io/tx/0xd183a55586addbce7a6c1e3849d495a8dca5a0c8d5f96e876d25dc50bd97c27a#eventlog
const TRANSFER_TOPIC: B256 =
    b256!("ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef");

sol!(
    #[sol(rpc)]
    IERC20,
    "src/abi/IERC20.json"
);

// We know that at least one WETH transfer hapenned at this block:
// https://etherscan.io/tx/0xd183a55586addbce7a6c1e3849d495a8dca5a0c8d5f96e876d25dc50bd97c27a
const BLOCK: u64 = 19633475;

#[tokio::main]
async fn main() -> Result<()> {
    dotenv().ok();
    let http_rpc = env::var("HTTP_RPC")?.parse()?;
    let provider = ProviderBuilder::new()
        .network::<AnyNetwork>()
        .on_http(http_rpc)?;
    let receipts = provider
        .get_block_receipts(BLOCK.into())
        .await?
        .expect(&format!("Block {BLOCK} should already exists"));

    for receipt in receipts {
        // LINE BELOW DOES NOT COMPILE:
        for log in &receipt.inner.inner.logs() {
            if log.address() == WETH {
                if log.topics()[0] == TRANSFER_TOPIC {
                    let transfer_log = Transfer::decode_log(log.as_ref(), true)?;
                    let transfer = transfer_log.data;
                    println!(
                        "{:?} has transfered {} ETH at transaction {:?} to {:?}",
                        transfer.src,
                        format_ether(transfer.wad),
                        receipt.transaction_hash,
                        transfer.dst
                    );
                }
            }
        }
    }

    Ok(())
}
```

I am trying to get the logs from a block receipt.

However I get the compilation error:
```
error[E0599]: no method named `logs` found for struct `alloy_consensus::receipt::any::AnyReceiptEnvelope<alloy_rpc_types::eth::log::Log>` in the current scope
  --> src/main.rs:40:42
   |
40 |         for log in &receipt.inner.inner.logs() {
   |                                         ^^^^ method not found in `AnyReceiptEnvelope<Log>`
   |
   = note: the method was found for
           - `alloy_consensus::receipt::any::AnyReceiptEnvelope`
```

But the `logs` method should be implemented [here](https://github.com/alloy-rs/alloy/blob/fe5754c06ecfcf563224ab6f60240e056cf99a3d/crates/consensus/src/receipt/any.rs#L59)

I think that the implementation is missing the generic parameters, that is why it is not accessible in my code:
```diff
+ impl<T> AnyReceiptEnvelope<T> {
- impl AnyReceiptEnvelope {
```

PS: In latest branch to still access the log we can do the following:
```diff
+ for log in &receipt.inner.inner.inner.receipt.logs {
- for log in &receipt.inner.inner.logs() {
```

But that is less user friendly than using the method.

--- a/crates/consensus/src/receipt/any.rs
+++ b/crates/consensus/src/receipt/any.rs
@@ -24,12 +24,10 @@ pub struct AnyReceiptEnvelope<T = Log> {
     pub r#type: u8,
 }
 
-impl AnyReceiptEnvelope {
-    /// Returns whether this is a legacy receipt (type 0)
-    pub const fn is_legacy(&self) -> bool {
-        self.r#type == 0
-    }
-
+impl<T> AnyReceiptEnvelope<T>
+where
+    T: Encodable,
+{
     /// Calculate the length of the rlp payload of the network encoded receipt.
     pub fn rlp_payload_length(&self) -> usize {
         let length = self.inner.length();
@@ -38,6 +36,13 @@ impl AnyReceiptEnvelope {
         } else {
             length + 1
         }
+    }
+}
+
+impl<T> AnyReceiptEnvelope<T> {
+    /// Returns whether this is a legacy receipt (type 0)
+    pub const fn is_legacy(&self) -> bool {
+        self.r#type == 0
     }
 
     /// Return true if the transaction was successful.
@@ -56,7 +61,7 @@ impl AnyReceiptEnvelope {
     }
 
     /// Return the receipt logs.
-    pub fn logs(&self) -> &[Log] {
+    pub fn logs(&self) -> &[T] {
         &self.inner.receipt.logs
     }
 

--- a/crates/consensus/src/receipt/receipts.rs
+++ b/crates/consensus/src/receipt/receipts.rs
@@ -1,3 +1,5 @@
+use core::borrow::Borrow;
+
 use super::TxReceipt;
 use alloy_primitives::{Bloom, Log};
 use alloy_rlp::{length_of_length, BufMut, Decodable, Encodable};
@@ -23,16 +25,19 @@ pub struct Receipt<T = Log> {
     pub logs: Vec<T>,
 }
 
-impl Receipt {
+impl<T> Receipt<T>
+where
+    T: Borrow<Log>,
+{
     /// Calculates [`Log`]'s bloom filter. this is slow operation and [ReceiptWithBloom] can
     /// be used to cache this value.
     pub fn bloom_slow(&self) -> Bloom {
-        self.logs.iter().collect()
+        self.logs.iter().map(Borrow::borrow).collect()
     }
 
     /// Calculates the bloom filter for the receipt and returns the [ReceiptWithBloom] container
     /// type.
-    pub fn with_bloom(self) -> ReceiptWithBloom {
+    pub fn with_bloom(self) -> ReceiptWithBloom<T> {
         self.into()
     }
 }
@@ -94,38 +99,17 @@ impl TxReceipt for ReceiptWithBloom {
     }
 }
 
-impl From<Receipt> for ReceiptWithBloom {
-    fn from(receipt: Receipt) -> Self {
+impl<T> From<Receipt<T>> for ReceiptWithBloom<T>
+where
+    T: Borrow<Log>,
+{
+    fn from(receipt: Receipt<T>) -> Self {
         let bloom = receipt.bloom_slow();
         ReceiptWithBloom { receipt, logs_bloom: bloom }
     }
 }
 
-impl ReceiptWithBloom {
-    /// Create new [ReceiptWithBloom]
-    pub const fn new(receipt: Receipt, bloom: Bloom) -> Self {
-        Self { receipt, logs_bloom: bloom }
-    }
-
-    /// Consume the structure, returning only the receipt
-    #[allow(clippy::missing_const_for_fn)] // false positive
-    pub fn into_receipt(self) -> Receipt {
-        self.receipt
-    }
-
-    /// Consume the structure, returning the receipt and the bloom filter
-    #[allow(clippy::missing_const_for_fn)] // false positive
-    pub fn into_components(self) -> (Receipt, Bloom) {
-        (self.receipt, self.logs_bloom)
-    }
-
-    fn payload_len(&self) -> usize {
-        self.receipt.status.length()
-            + self.receipt.cumulative_gas_used.length()
-            + self.logs_bloom.length()
-            + self.receipt.logs.length()
-    }
-
+impl<T: Encodable> ReceiptWithBloom<T> {
     /// Returns the rlp header for the receipt payload.
     fn receipt_rlp_header(&self) -> alloy_rlp::Header {
         alloy_rlp::Header { list: true, payload_length: self.payload_len() }
@@ -140,8 +124,37 @@ impl ReceiptWithBloom {
         self.receipt.logs.encode(out);
     }
 
+    fn payload_len(&self) -> usize {
+        self.receipt.status.length()
+            + self.receipt.cumulative_gas_used.length()
+            + self.logs_bloom.length()
+            + self.receipt.logs.length()
+    }
+}
+
+impl<T> ReceiptWithBloom<T> {
+    /// Create new [ReceiptWithBloom]
+    pub const fn new(receipt: Receipt<T>, bloom: Bloom) -> Self {
+        Self { receipt, logs_bloom: bloom }
+    }
+
+    /// Consume the structure, returning only the receipt
+    #[allow(clippy::missing_const_for_fn)] // false positive
+    pub fn into_receipt(self) -> Receipt<T> {
+        self.receipt
+    }
+
+    /// Consume the structure, returning the receipt and the bloom filter
+    #[allow(clippy::missing_const_for_fn)] // false positive
+    pub fn into_components(self) -> (Receipt<T>, Bloom) {
+        (self.receipt, self.logs_bloom)
+    }
+
     /// Decodes the receipt payload
-    fn decode_receipt(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+    fn decode_receipt(buf: &mut &[u8]) -> alloy_rlp::Result<Self>
+    where
+        T: Decodable,
+    {
         let b: &mut &[u8] = &mut &**buf;
         let rlp_head = alloy_rlp::Header::decode(b)?;
         if !rlp_head.list {
@@ -169,7 +182,7 @@ impl ReceiptWithBloom {
     }
 }
 
-impl alloy_rlp::Encodable for ReceiptWithBloom {
+impl<T: Encodable> Encodable for ReceiptWithBloom<T> {
     fn encode(&self, out: &mut dyn BufMut) {
         self.encode_fields(out);
     }
@@ -183,7 +196,7 @@ impl alloy_rlp::Encodable for ReceiptWithBloom {
     }
 }
 
-impl alloy_rlp::Decodable for ReceiptWithBloom {
+impl<T: Decodable> Decodable for ReceiptWithBloom<T> {
     fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
         Self::decode_receipt(buf)
     }

--- a/crates/rpc-types/src/eth/log.rs
+++ b/crates/rpc-types/src/eth/log.rs
@@ -78,17 +78,34 @@ impl Log<LogData> {
     }
 }
 
+impl<T> alloy_rlp::Encodable for Log<T>
+where
+    for<'a> &'a T: Into<LogData>,
+{
+    fn encode(&self, out: &mut dyn alloy_rlp::BufMut) {
+        self.reserialize_inner().encode(out)
+    }
+
+    fn length(&self) -> usize {
+        self.reserialize_inner().length()
+    }
+}
+
 impl<T> Log<T>
 where
     for<'a> &'a T: Into<LogData>,
 {
-    /// Reserialize the data.
+    /// Reserialize the inner data, returning an [`alloy_primitives::Log`].
+    pub fn reserialize_inner(&self) -> alloy_primitives::Log {
+        alloy_primitives::Log { address: self.inner.address, data: (&self.inner.data).into() }
+    }
+
+    /// Reserialize the data, returning a new `Log` object wrapping an
+    /// [`alloy_primitives::Log`]. this copies the log metadata, preserving
+    /// the original object.
     pub fn reserialize(&self) -> Log<LogData> {
         Log {
-            inner: alloy_primitives::Log {
-                address: self.inner.address,
-                data: (&self.inner.data).into(),
-            },
+            inner: self.reserialize_inner(),
             block_hash: self.block_hash,
             block_number: self.block_number,
             block_timestamp: self.block_timestamp,

